pkg/lib/operation/project/remote/file/download/operation.go

```diff
@@ -24,206 +31,184 @@ type dependencies interface {
 	Telemetry() telemetry.Telemetry
 }
 
-type Options struct {
-	File        *keboola.FileDownloadCredentials
-	Output      string
-	AllowSliced bool
+type downloader struct {
+	dependencies
+	options Options
+	bar     *progressbar.ProgressBar
+	slices  []string
 }
 
-func (o *Options) ToStdOut() bool {
-	return o.Output == "-"
+func Run(ctx context.Context, o Options, d dependencies) (returnErr error) {
+	return (&downloader{options: o, dependencies: d}).Download(ctx)
 }
 
-// GetOutput returns an `io.WriteCloser` which either wraps a file or stdout.
-//
-// If `o.Output` is "-", then it wraps stdout, and `Close` is a no-op.
-// Otherwise wraps a file.
-//
-// If `file` is not an empty string, then `o.Output` is treated as a directory
-// and the file created will be at `path.Join(o.Output, file)`.
-func (o *Options) GetOutput(file string) (io.WriteCloser, error) {
-	if o.Output == "-" {
-		return &outputWriter{}, nil
-	} else {
-		var output string
-		if len(file) > 0 {
-			output = path.Join(o.Output, file)
-		} else {
-			output = o.Output
-		}
-
-		file, err := os.Create(output) // nolint: forbidigo
-		if err != nil {
-			return nil, errors.Errorf(`cannot create file "%s": %w`, output, err)
-		}
-		return &outputWriter{file}, nil
-	}
-}
-
-func Run(ctx context.Context, opts Options, d dependencies) (err error) {
+func (d *downloader) Download(ctx context.Context) (returnErr error) {
 	ctx, span := d.Telemetry().Tracer().Start(ctx, "keboola.go.operation.project.remote.file.download")
-	defer span.End(&err)
+	defer span.End(&returnErr)
 
-	if opts.File.IsSliced {
-		if !opts.AllowSliced {
-			err = runDownloadForceUnsliced(ctx, &opts, d.Logger())
-			if err != nil {
-				return err
-			}
-		} else {
-			err = runDownloadSliced(ctx, &opts)
-			if err != nil {
-				return err
+	// Log start and end
+	if !d.options.ToStdout() {
+		defer func() {
+			if returnErr == nil {
+				d.Logger().Infof(`File "%d" downloaded to "%s".`, d.options.File.ID, d.options.FormattedOutput())
 			}
-		}
-	} else {
-		err = runDownloadWhole(ctx, &opts)
-		if err != nil {
-			return err
-		}
+		}()
 	}
 
-	if !opts.ToStdOut() {
-		d.Logger().Infof(`File "%d" downloaded to "%s".`, opts.File.ID, opts.Output)
+	// Get slices
+	if slices, err := d.getSlices(ctx); err == nil {
+		d.slices = slices
+	} else {
+		return err
 	}
 
-	return nil
-}
-
-// download slices to `opts.Output` as one file.
-func runDownloadForceUnsliced(ctx context.Context, opts *Options, logger log.Logger) error {
-	logger.Infof(`Creating file "%s"`, opts.Output)
-	dst, err := opts.GetOutput("")
+	// Get total size
+	size, err := d.totalSize(ctx)
 	if err != nil {
 		return err
 	}
+
+	// Create progress bar, it writes to stderr
+	d.bar = progressbar.DefaultBytes(size, "Downloading")
 	defer func() {
-		err = dst.Close()
+		if closeErr := d.bar.Close(); returnErr == nil && closeErr != nil {
+			returnErr = closeErr
+		}
```

Comment:
Na viacerych miestach tu chybalo volanie `Close`, na tomto mieste sa ukoncuje progress bar.
Bez toho to mohlo skoncit na `99%`, aj ked sa subor stiahol cely, no bar sa nestihol vykreslit.